src/python_bindings/py_data_stats.cpp

```diff
@@ -0,0 +1,162 @@
+#include "py_data_stats.h"
+
+#include "statistics/data_stats.h"
+
+namespace python_bindings {
+
+namespace py = pybind11;
+namespace mo = model;
+
+size_t PyDataStats::GetNumberOfValues(size_t index) const {
+    return GetAlgorithm<algos::DataStats>(algorithm_).NumberOfValues(index);
+}
+
+size_t PyDataStats::GetNumberOfColumns() const {
+    return GetAlgorithm<algos::DataStats>(algorithm_).GetNumberOfColumns();
+}
+
+std::vector<size_t> PyDataStats::GetColumnsWithNull() const {
+    return GetAlgorithm<algos::DataStats>(algorithm_).GetColumnsWithNull();
+}
+
+std::vector<size_t> PyDataStats::GetColumnsWithUniqueValues() const {
+    return dynamic_cast<algos::DataStats&>(*algorithm_).GetColumnsWithUniqueValues();
+}
+
+size_t PyDataStats::Distinct(size_t index) const {
+    return dynamic_cast<algos::DataStats&>(*algorithm_).Distinct(index);
+}
+
+bool PyDataStats::IsCategorical(size_t index, size_t quantity) {
+    return dynamic_cast<algos::DataStats&>(*algorithm_).IsCategorical(index, quantity);
+}
+
+std::vector<std::vector<std::string>> PyDataStats::ShowSample(size_t start_row, size_t end_row,
+                                                              size_t start_col, size_t end_col,
+                                                              size_t str_len, size_t unsigned_len,
+                                                              size_t double_len) const {
+    const algos::DataStats& stat = GetAlgorithm<algos::DataStats>(algorithm_);
+    return stat.ShowSample(start_row, end_row, start_col, end_col, str_len, unsigned_len,
+                           double_len);
+}
+
+py::object GetValue(algos::Statistic stat) {
+    const std::byte* data = stat.GetData();
+    const mo::Type* type = stat.GetType();
+
+    if (data == nullptr) return py::none();
+
+    switch (type->GetTypeId()) {
+        case mo::TypeId::kDouble:
+            return py::float_(mo::Type::GetValue<mo::Double>(data));
+        case mo::TypeId::kInt:
+            return py::int_(mo::Type::GetValue<mo::Int>(data));
+        case mo::TypeId::kString:
+            return py::str(mo::Type::GetValue<mo::String>(data));
+        default:
+            assert(false);
+            __builtin_unreachable();
+    }
+}
+
+py::object PyDataStats::GetAverage(size_t index) const {
+    algos::Statistic stat = GetAlgorithm<algos::DataStats>(algorithm_).GetAvg(index);
+    return GetValue(stat);
```

Comment:
You should make a function template for this and use it everywhere applicable.

Reply:
Don't quite understand what function templates I should implement

Reply:
I doubt this one would look good either, don't bother.